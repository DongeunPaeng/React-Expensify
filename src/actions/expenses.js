import db from "../firebase/firebase";

// ADD_EXPENSE
export const addExpense = expense => ({
  type: "ADD_EXPENSE",
  expense
});

// ASYNC ACTION (that returns a function that dispatches another function that returns the final object)
export const startAddExpense = (expenseData = {}) => {
  return async dispatch => {
    const {
      description = "",
      notes = "",
      amount = 0,
      createdAt = 0
    } = expenseData;

    // ADD EXPENSE TO FIRESTORE
    const expense = { description, notes, amount, createdAt };

    await db
      .collection("expenses")
      .add(expense)
      .then(doc => {
        dispatch(
          addExpense({
            id: doc.id,
            ...expense
          })
        );
      });
  };
};

// REMOVE_EXPENSE
export const removeExpense = id => ({
  type: "REMOVE_EXPENSE",
  id
});

export const startRemoveExpense = id => {
  return async dispatch => {
    await db
      .collection("expenses")
      .doc(id)
      .delete()
      .then()
      .catch(err => console.log("err happened deleting a doc...", err));
    dispatch(removeExpense(id));
  };
};

// EDIT_EXPENSE
export const editExpense = (id, updates) => ({
  type: "EDIT_EXPENSE",
  id,
  updates
});

export const setExpenses = expenses => ({
  type: "SET_EXPENSES",
  expenses
});

export const startSetExpenses = () => {
  return async dispatch => {
    let expenses = [];
    await db
      .collection("expenses")
      .get()
      .then(snapshot => {
        snapshot.forEach(doc => {
          expenses.push({
            id: doc.id,
            ...doc.data()
          });
        });
      })
      .catch(err => {
        console.log(err);
      });
    dispatch(setExpenses(expenses));
  };
};
