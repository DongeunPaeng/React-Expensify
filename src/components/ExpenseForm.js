import React from "react";
import { DatePicker } from "rsuite";
import { DateTime } from "luxon";

export default class ExpenseForm extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      description: props.expense.description ? props.expense.description : "",
      note: props.expense.note ? props.expense.note : "",
      amount: props.expense.amount ? props.expense.amount : "",
      createdAt: props.expense.createdAt
        ? DateTime.fromMillis(props.expense.createdAt).ts
        : DateTime.local(),
      message: undefined
    };
  }

  onDescriptionChange = e => {
    const description = e.target.value;
    this.setState(() => ({ description }));
  };

  onNoteChange = e => {
    const note = e.target.value;
    this.setState(() => ({ note }));
  };

  onAmountChange = e => {
    const amount = e.target.value;
    this.setState(() => ({ amount }));
  };

  onDateChange = date => {
    const createdAt = DateTime.fromISO(date.toISOString()).ts;
    if (createdAt) {
      this.setState(() => ({ createdAt }));
    }
  };

  onSubmit = e => {
    e.preventDefault();
    if (!this.state.description || !this.state.amount) {
      this.setState(() => ({
        message: "please provide description and amount"
      }));
    } else {
      this.setState(() => ({ message: "" }));
      this.props.onSubmit({
        description: this.state.description,
        amount: parseInt(this.state.amount),
        createdAt: this.state.createdAt,
        notes: this.state.note
      });
    }
  };

  render() {
    return (
      <div>
        {this.state.message && <p>{this.state.message}</p>}
        <form onSubmit={this.onSubmit}>
          <input
            type="text"
            placeholder="Description"
            autoFocus
            value={this.state.description}
            onChange={this.onDescriptionChange}
          />
          <input
            type="number"
            placeholder="Amount"
            value={this.state.amount}
            onChange={this.onAmountChange}
          />
          <DatePicker onChange={this.onDateChange} />
          <textarea
            placeholder="type something(optional)"
            value={this.state.note}
            onChange={this.onNoteChange}
          ></textarea>
          <button>Add Expense</button>
        </form>
      </div>
    );
  }
}
