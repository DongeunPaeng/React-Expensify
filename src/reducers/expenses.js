const expensesIntialState = [];

export default (state = expensesIntialState, action) => {
  switch (action.type) {
    case "ADD_EXPENSE":
      return [...state, action.expense];
    case "REMOVE_EXPENSE":
      return state.filter(({ id }) => id !== action.id);
    case "EDIT_EXPENSE":
      return state.map(expense => {
        if (expense.id === action.id) {
          return {
            ...expense,
            ...action.updates
          };
        } else {
          return expense;
        }
      });
    case "SET_EXPENSES":
      return action.expenses;
    default:
      return state;
  }
};

// reducers do this thing: they receive previous state and actions to return the updated state.
