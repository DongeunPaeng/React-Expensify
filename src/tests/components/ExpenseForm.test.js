import React from "react";
import { shallow } from "enzyme";
import ExpenseForm from "../../components/ExpenseForm";
import expenses from "../fixtures/expenses";
import { DateTime } from "luxon";

test("should render ExpenseForm correctly", () => {
  const wrapper = shallow(<ExpenseForm />);
  expect(wrapper).toMatchSnapshot();
});

test("should render ExpenseForm with expense data", () => {
  const wrapper = shallow(<ExpenseForm expense={expenses[0]} />);
  expect(wrapper).toMatchSnapshot();
});

test("should render error for invalid form submission", () => {
  const wrapper = shallow(<ExpenseForm />);
  wrapper.find("form").simulate("submit", {
    preventDefault: () => {}
  });
  expect(wrapper.state("message").length).toBeGreaterThan(0);
  expect(wrapper).toMatchSnapshot();
});

test("should set description on input change", () => {
  const value = "New description";
  const wrapper = shallow(<ExpenseForm />);
  wrapper
    .find("input")
    .at(0)
    .simulate("change", {
      target: { value }
    });
  expect(wrapper.state("description")).toBe(value);
});

test("should set note on textarea change", () => {
  const value = "new notes";
  const wrapper = shallow(<ExpenseForm />);
  wrapper
    .find("textarea")
    .at(0)
    .simulate("change", {
      target: { value }
    });
  expect(wrapper.state("note")).toBe(value);
});

test("should set amount", () => {
  const value = 2350;
  const wrapper = shallow(<ExpenseForm />);
  wrapper
    .find("input")
    .at(1)
    .simulate("change", {
      target: { value }
    });
  expect(wrapper.state("amount")).toBe(value);
});

test("should call onSubmit prop for valid form submission", () => {
  const onSubmitSpy = jest.fn();
  const wrapper = shallow(
    <ExpenseForm expense={expenses[0]} onSubmit={onSubmitSpy} />
  );
  wrapper.find("form").simulate("submit", {
    preventDefault: () => {}
  });
  expect(wrapper.state("message")).toBe("");
  expect(onSubmitSpy).toHaveBeenLastCalledWith({
    description: expenses[0].description,
    amount: expenses[0].amount,
    notes: expenses[0].note || "",
    createdAt: DateTime.fromMillis(expenses[0].createdAt).ts
  });
});

test("should set new date on date change", () => {
  const now = new Date();
  const wrapper = shallow(<ExpenseForm />);
  wrapper.find("withLocale(defaultProps(DatePicker))").prop("onChange")(now);
  expect(wrapper.state("createdAt")).toEqual(
    DateTime.fromISO(now.toISOString()).ts
  );
});
