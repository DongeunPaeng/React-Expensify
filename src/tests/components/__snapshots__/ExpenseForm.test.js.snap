// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpenseForm correctly 1`] = `
<form
  className="form"
  onSubmit={[Function]}
>
  <input
    autoFocus={true}
    className="text-input"
    onChange={[Function]}
    placeholder="Description"
    type="text"
    value=""
  />
  <input
    className="text-input"
    onChange={[Function]}
    placeholder="Amount"
    type="number"
    value=""
  />
  <withLocale(defaultProps(DatePicker))
    onChange={[Function]}
  />
  <textarea
    className="textarea"
    onChange={[Function]}
    placeholder="type something(optional)"
    value=""
  />
  <div>
    <button
      className="button"
    >
      Save Expense
    </button>
  </div>
</form>
`;

exports[`should render ExpenseForm with expense data 1`] = `
<form
  className="form"
  onSubmit={[Function]}
>
  <input
    autoFocus={true}
    className="text-input"
    onChange={[Function]}
    placeholder="Description"
    type="text"
    value="test01"
  />
  <input
    className="text-input"
    onChange={[Function]}
    placeholder="Amount"
    type="number"
    value={195}
  />
  <withLocale(defaultProps(DatePicker))
    onChange={[Function]}
  />
  <textarea
    className="textarea"
    onChange={[Function]}
    placeholder="type something(optional)"
    value=""
  />
  <div>
    <button
      className="button"
    >
      Save Expense
    </button>
  </div>
</form>
`;

exports[`should render error for invalid form submission 1`] = `
<form
  className="form"
  onSubmit={[Function]}
>
  <p
    className="form__message"
  >
    Please provide description and amount.
  </p>
  <input
    autoFocus={true}
    className="text-input"
    onChange={[Function]}
    placeholder="Description"
    type="text"
    value=""
  />
  <input
    className="text-input"
    onChange={[Function]}
    placeholder="Amount"
    type="number"
    value=""
  />
  <withLocale(defaultProps(DatePicker))
    onChange={[Function]}
  />
  <textarea
    className="textarea"
    onChange={[Function]}
    placeholder="type something(optional)"
    value=""
  />
  <div>
    <button
      className="button"
    >
      Save Expense
    </button>
  </div>
</form>
`;
